crates/criticalup-core/src/download_server_client.rs

```diff
@@ -126,42 +126,41 @@ impl DownloadServerClient {
         let mut current_retries: u8 = 0;
         let mut current_retry_backoff = CLIENT_RETRY_BACKOFF;
 
-        // This `try_clone()` will be `None` if request body is a stream.
-        let req_outer = req.try_clone().ok_or(Error::RequestCloningFailed)?;
-        let mut response_result = self.client.execute(req_outer);
+        while current_retries < CLIENT_MAX_RETRIES {
+            // This `try_clone()` will be `None` if request body is a stream.
+            let response_result = self
+                .client
+                .execute(req.try_clone().ok_or(Error::RequestCloningFailed)?);
 
-        while current_retries < CLIENT_MAX_RETRIES && response_result.is_err() {
-            let req = req.try_clone().ok_or(Error::RequestCloningFailed)?;
-
-            response_result = self.client.execute(req);
             if response_result.is_ok() {
-                break;
+                let response = response_result?;
+
+                return Err(self.err_from_response(
```

Comment:
This is kinda confusing, you have a `return Err(...)` then inside of it a `return Ok(...)`.

Reply:
Yea I was trying to preserve what @pietroalbini had.